File: backend/go-chi-hive/internal/cells/user/ports.go
package user

import (
	"context"
)

type RegisterCommand struct {
	Username string
	Email    string
	Password string
}

type LoginCommand struct {
	Email    string
	Password string
}

type UpdateUserCommand struct {
	ID       int64
	Username *string
	Email    *string
	Password *string
	Bio      *string
	Image    *string
}

// Service defines the inbound port for user-related use cases.
type Service interface {
	Register(ctx context.Context, cmd RegisterCommand) (*User, error)
	Login(ctx context.Context, cmd LoginCommand) (*User, error)
	GetUser(ctx context.Context, id int64) (*User, error)
	UpdateUser(ctx context.Context, cmd UpdateUserCommand) (*User, error)

	GetProfile(ctx context.Context, username string, observerID *int64) (*Profile, error)
	FollowUser(ctx context.Context, followerID int64, username string) (*Profile, error)
	UnfollowUser(ctx context.Context, followerID int64, username string) (*Profile, error)
}

// Repository defines the outbound port for user data persistence.
type Repository interface {
	Create(ctx context.Context, user *User) error
	FindByEmail(ctx context.Context, email string) (*User, error)
	FindByUsername(ctx context.Context, username string) (*User, error)
	FindByID(ctx context.Context, id int64) (*User, error)
	Update(ctx context.Context, user *User) error

	GetProfileByUsername(ctx context.Context, username string, observerID *int64) (*Profile, error)
	Follow(ctx context.Context, followerID, followedID int64) error
	Unfollow(ctx context.Context, followerID, followedID int64) error
}

// PasswordHasher defines the outbound port for password security.
type PasswordHasher interface {
	Hash(password string) (string, error)
	Compare(hashedPassword, password string) error
}

// TokenGenerator defines the outbound port for generating tokens.
type TokenGenerator interface {
	Generate(userID int64) (string, error)
}

File: backend/typescript-node-express-hive/src/cells/user/ports.ts
import { User, type Profile } from "./user.js";

export interface RegisterCommand {
    username: string;
    email: string;
    password: string;
}

export interface LoginCommand {
    email: string;
    password: string;
}

export interface UpdateUserCommand {
    id: number;
    username?: string;
    email?: string;
    password?: string;
    bio?: string;
    image?: string | null;
}

export interface UserService {
    register(cmd: RegisterCommand): Promise<User>;
    login(cmd: LoginCommand): Promise<User>;
    getUser(id: number): Promise<User>;
    getUserByUsername(username: string): Promise<User>;
    updateUser(cmd: UpdateUserCommand): Promise<User>;

    getProfile(username: string, observerId?: number): Promise<Profile>;
    followUser(followerId: number, username: string): Promise<Profile>;
    unfollowUser(followerId: number, username: string): Promise<Profile>;
}

export interface UserRepository {
    create(user: User): Promise<void>;
    findByEmail(email: string): Promise<User | null>;
    findByUsername(username: string): Promise<User | null>;
    findById(id: number): Promise<User | null>;
    update(user: User): Promise<void>;

    getProfileByUsername(username: string, observerId?: number): Promise<Profile | null>;
    follow(followerId: number, followedId: number): Promise<void>;
    unfollow(followerId: number, followedId: number): Promise<void>;
}
